docs/src/main/asciidoc/picocli.adoc

```diff
@@ -0,0 +1,178 @@
+
+@CommandLine.Command // <1>
+public class HelloCommand implements Runnable {
+
+    @CommandLine.Option(names = {"-n", "--name"}, description = "Who will we greet?", defaultValue = "World")
+    String name;
+
+    private final GreetingService greetingService;
+
+    @Inject // <2>
+    public HelloCommand(GreetingService greetingService) {
+        this.greetingService = greetingService;
+    }
+
+    @Override
+    public void run() {
+        greetingService.sayHello(name);
+    }
+}
+
+@Dependent
+class GreetingService {
+    void sayHello(String name) {
+        System.out.println("Hello " + name + "!");
+    }
+}
+----
+<1> If there is only one class annotated with `picocli.CommandLine.Command` it will be used as entry point to Picocli CommandLine.
+<2> All classes annotated with `picocli.CommandLine.Command` are register as beans in CDI.
+
+== Picocli Application with multiple Commands
+
+When multiple classes has `picocli.CommandLine.Command` annotation, then one of them need to be also annotated with `io.quarkus.picocli.runtime.annotations.TopCommand`.
+This can be overwritten with `quarkus.picocli.top-command` property.
+
+[source,java]
+----
+package com.acme.picocli;
+
+import io.quarkus.picocli.runtime.annotations.TopCommand;
+import picocli.CommandLine;
+
+@TopCommand
+@CommandLine.Command(mixinStandardHelpOptions = true, subcommands = {HelloCommand.class, GoodByeCommand.class})
+public class EntryCommand {
+}
+
+@CommandLine.Command(name = "hello", description = "Greet World!")
+class HelloCommand implements Runnable {
+
+    @Override
+    public void run() {
+        System.out.println("Hello World!");
+    }
+}
+
+@CommandLine.Command(name = "goodbye", description = "Say goodbye to World!")
+class GoodByeCommand implements Runnable {
+
+    @Override
+    public void run() {
+        System.out.println("Goodbye World!");
+    }
+}
+----
+
+== Customizing Picocli CommandLine instance
+
+You can customize CommandLine classes used by `picocli` extension by producing your own bean instance:
+
+[source,java]
+----
+package com.acme.picocli;
+
+import io.quarkus.picocli.runtime.PicocliCommandLineFactory;
+import io.quarkus.picocli.runtime.annotations.TopCommand;
+import picocli.CommandLine;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+
+@TopCommand
+@CommandLine.Command
+public class EntryCommand implements Runnable {
+    @CommandLine.Spec
+    CommandLine.Model.CommandSpec spec;
+
+    @Override
+    public void run() {
+        System.out.println("My name is: " + spec.name());
+    }
+}
+
+@ApplicationScoped
+class CustomConfiguration {
+
+    @Produces
+    CommandLine customCommandLine(PicocliCommandLineFactory factory) { // <1>
+        return factory.create().setCommandName("CustomizedName");
+    }
+}
+----
+<1> `PicocliCommandLineFactory` will create an instance of CommandLine with `TopCommand` and `CommandLine.IFactory` injected.
+
+== Native mode support
```

Comment:
Can't we generate this configuration as part of the Quarkus build? I reckon there might be some redundancy but that's the Quarkus way of doing things.

Reply:
Do you mean that we should have property like `quarkus.picocli.command-name` which would configure this? I can expose most of setters from CommandLine as properties, but there are many, optional, things which can be configured there only programmatically. So I would like to leave `PicocliCommandLineFactory` (maybe with different name) to allow easy reconfiguration of CommandLine. 

I'm also not sure which of CommandLine properties should be configured in runtime. This maybe not a real issue, but for example setting `QUARKUS_PICOCLI_TRIM_QUOTES` in env will change behavior of your command line application, e.g. it will start to reacting to arguments in different way.

Reply:
No my comment was about the native mode. We shouldn't need an annotation processor to configure the native mode, everything should be done in the Quarkus processor via build items. We can for sure help with this!

Reply:
I will try to port picocli-codegen code to build steps. Reflection and resource bundles should be straightforward. Not sure about dynamic proxies, but I think I saw code for this also in quarkus.